metrics/prometheus.go

```diff
@@ -0,0 +1,133 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cloudwan/gohan/util"
+	prometheusmetrics "github.com/deathowl/go-metrics-prometheus"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	metrics "github.com/rcrowley/go-metrics"
+)
+
+const configPrefix = "metrics/prometheus/"
+
+func getString(config *util.Config, key, defaultValue string) string {
+	return config.GetString(configPrefix+key, defaultValue)
+}
+
+func getDuration(config *util.Config, key string, defaultValue time.Duration) time.Duration {
+	return config.GetDuration(configPrefix+key, defaultValue)
+}
+
+type prometheusConfig struct {
+	flushInterval time.Duration
+	timerBuckets  []float64
+	namespace     string
+	subsystem     string
+	serverAddress string
+	serverPath    string
+	serverBackoff time.Duration
+}
+
+type prometheusExporter struct {
+	provider *prometheusmetrics.PrometheusConfig
+	config   prometheusConfig
+}
+
+func (pe *prometheusExporter) Setup(config *util.Config, mainAddress string) error {
+	if err := pe.setServerAddress(config, mainAddress); err != nil {
+		return err
+	}
+
+	pe.config.flushInterval = getDuration(config, "flush_interval", 10*time.Second)
+	pe.config.namespace = getString(config, "namespace", "gohan")
+	pe.config.subsystem = getString(config, "subsystem", "")
+
+	var err error
+	pe.config.timerBuckets, err = getPercentilesFrom(config, configPrefix+"timer_buckets")
+	if err != nil {
+		return err
+	}
+
+	pe.provider = prometheusmetrics.NewPrometheusProvider(metrics.DefaultRegistry, pe.config.namespace, pe.config.subsystem,
+		prometheus.DefaultRegisterer, pe.config.flushInterval)
+	pe.provider.WithTimerBuckets(pe.config.timerBuckets)
+
+	pe.config.serverPath = getString(config, "path", "/metrics")
+	pe.config.serverBackoff = getDuration(config, "backoff", 5*time.Second)
+
+	return nil
+}
+
+func (pe *prometheusExporter) setServerAddress(config *util.Config, mainAddress string) error {
+	mainSplit := strings.Split(mainAddress, ":")
+	if len(mainSplit) != 2 {
+		return errors.Errorf("Incorrect main gohan address '%s', cannot deduct prometheus metrics address", mainAddress)
+	}
+	mainPort, err := strconv.Atoi(mainSplit[1])
+	if err != nil {
+		return errors.Errorf("Incorrect main gohan address '%s', port is not a number", mainAddress)
+	}
+
+	defaultAddress := fmt.Sprintf("%s:%d", mainSplit[0], mainPort+2) // healthcheck uses +1
```

Comment:
maybe lets create an enum with port offsets, to avoid magic numbers and comments

Reply:
Done.

Reply:
It doesn't look done to me